main.py

```diff
@@ -0,0 +1,1605 @@
+                topic_config.remove_option(account_type, key)
+        with open('saved_topics.ini', 'w') as topic_configfile:
+            topic_config.write(topic_configfile)
+
+def do_subscribe(): #client.subscribe() work in here
+    global topic
+    global tab1_subscribed_list
+
+    if client_flag == 0:
+        terminal_print('Please select a target device.')
+        return 
+    topic = tab1_sub_to_topic.get() #grab what was typed in the sub_to_topic entry box
+    address = topics.compare_subs(topic, mqtt_topic_prefix, target_device) #function to compare topics to get their actual address if selected from listbox
+    terminal_print("Address: " + address)
+    client.subscribe(address, qos=0)
+
+    #check if it's in our list of subscribed topics
+    if address not in subscribed_topics_list:   #make sure there's no duplicates
+        subscribed_topics_list.append(address)  #add to our subscribed topics list
+
+        topic_config[account_type][target_device] = {address}   #and add to file
+        with open('saved_topics.ini', 'w') as topic_configfile:
+            topic_config.write(topic_configfile)
+        
+def check_subscribed_topics():
+    file_exists = os.path.exists('saved_topics.ini')
+    if file_exists is not True:  #create file and add section for subscribe topics
+        topic_config['Prod'] = {}
+        topic_config['Beta'] = {}
+        topic_config['Dev'] = {}
+        topic_config['Feat'] = {}
+
+        with open('saved_topics.ini', 'w') as topic_configfile:
+            topic_config.write(topic_configfile)
+
+    else:   #parse existing file
+        topic_config.read('saved_topics.ini')
+        subscribed_topics = topic_config[account_type]
+
+        if subscribed_topics == '': #nothing to subscribe to, move on
+            return
+        else:
+            for value in topic_config[account_type]:  #for loop to send each topic to subscribe function
+                subscribed_topics_list.append(value)  #also store the topic in a list 
+                auto_subscribe(value)
+                print('subscribed list:', subscribed_topics_list)
+
+def pub_message_check(msg): #user custom message
+    if msg == '':
+        msg = None
+    else:
+        msg = msg
+    return msg
+
+def check_message(message, curr_msg_topic):  #function to handle exceptions
+    #print("Raw data: ", message)
+    #print(type(message))
+    try:
+        unpacked_json = json.loads(message)
+    except Exception as e:
+        print("Couldn't parse raw data: %s" % message, e)
+    else:
+        #print("JSON:", unpacked_json)
+        #print(type(unpacked_json))
+        sort_message(unpacked_json, curr_msg_topic)
+
+def sort_message(unpacked_json, curr_msg_topic):       
+    message_array = []
+    
+    '''
+    The for-loop below used to work without issues, but now it's giving an error depending on
+    which message we receive. 
+    Error msg: "Caught exception in on_message: 'list' object has no attribute 'items'"
```

Comment:
What did you see in check_message() when your print statements right before the call to sort_message() are not commented out, when the for loop gets an exception?

Reply:
Un-commented the for-loop inside that for-loop and it's fine now.... for now

Reply:
I have a hunch about what this might be --

The root object of a JSON string can either be a list, or an object, which would then be unpacked as a list or a dictionary in Python respectively

For instance:

`"[1, 2, 3]"` will unpack as a list

`"{'a':1, 'b':2, 'c':3}"` will unpack as a dictionary

so the key,value iterator will work when the JSON string provided is a dictionary, but if the JSON string provided is a list, then you will get this error

Our backend allows FW devices to upload either single device messages as objects (`"{device message}"`), or multiple device messages as a list (`[{device message}, {device message}, {device message}]`).

If I had to guess, it's those bulk/list-type messages that are causing errors with this code. If you confirm that is the case, the solution would be to check whether what you unpacked is a list before passing it to sort_message, and if it is, call sort_message once for each element of the list